### PR TITLE
Ix intp

### PR DESCRIPTION
**Pull request by jaimefrio**

### CODE REVIEW

File: numpy/lib/index_tricks.py [hunk 1]
```diff
@@ -71,17 +71,15 @@ def ix_(*args):
     """
     out = []
     nd = len(args)
-    baseshape = [1]*nd
-    for k in range(nd):
-        new = _nx.asarray(args[k])
+    for k, new in enumerate(args):
+        # Explicitly type empty sequences to avoid float default
+        new = asarray(new, dtype=None if new else _nx.intp)
```
Comment (charris) [inline numpy/lib/index_tricks.py]: `if new` won't work if `new` is an array of len > 1.

**Comment by jaimefrio** [comment]
Pushed new commit, ready for re-review.

File: numpy/lib/index_tricks.py [hunk 2]
```diff
@@ -71,17 +71,17 @@ def ix_(*args):
...
-    baseshape = [1]*nd
-    for k in range(nd):
-        new = _nx.asarray(args[k])
-        if (new.ndim != 1):
+    for k, new in enumerate(args):
+        new = asarray(new)
+        if new.size == 0:
+            # Explicitly type empty arrays to avoid float default
+            new = new.astype(_nx.intp)
+        if new.ndim != 1:
```
Comment (charris) [inline numpy/lib/index_tricks.py]: This check would be better if done first before the rest.

**Comment by charris** [comment]
Just needs a little bit more...

**Comment by jaimefrio** [comment]
Reordered the checks, should be ready to go.

**Comment by charris** [comment]
Merged. Thanks Jaime.

File: numpy/lib/tests/test_index_tricks.py [hunk 3]
```diff
@@ -169,6 +169,34 @@ def test_simple_1(self):
         assert_equal(a[:, :3, [1, 2]], a[s_[:, :3, [1, 2]]])
 
 
+class TestIx_(TestCase):
+    def test_regression_1(self):
+        # Test empty inputs create ouputs of indexing type, gh-5804
+        # Test both lists and arrays
+        for func in (range, np.arange):
```
Comment (eric-wieser) [inline numpy/lib/tests/test_index_tricks.py]: Wait, why is it desirable that `np.arange(0, dtype=float)` is converted to `dtype=intp`? If the user is explicit about what they asked for, it seems wrong to override them

**Pull request merged**